xcm/xcm-executor/src/lib.rs

```diff
@@ -16,142 +16,237 @@
-		//   including depositing fees for effects from holding account.
-
-		for effect in effects.into_iter() {
-			let _ = Self::execute_effects(&origin, &mut holding, effect)?;
+		if let Some((mut holding, effects)) = maybe_holding_effects {
+			for effect in effects.into_iter() {
+				total_surplus += Self::execute_effects(&origin, &mut holding, effect, trader)?;
+			}
 		}
 
-		// TODO: stuff that should happen after effects including refunding unused fees.
-
-		Ok(())
-	}
-}
-
-impl<Config: config::Config> XcmExecutor<Config> {
-	fn reanchored(mut assets: Assets, dest: &MultiLocation) -> Vec<MultiAsset> {
-		let inv_dest = Config::LocationInverter::invert_location(&dest);
-		assets.reanchor(&inv_dest);
-		assets.into_assets_iter().collect::<Vec<_>>()
+		Ok(total_surplus)
 	}
 
-	fn execute_effects(_origin: &MultiLocation, holding: &mut Assets, effect: Order) -> XcmResult {
+	fn execute_effects(
+		origin: &MultiLocation,
+		holding: &mut Assets,
+		effect: Order<Config::Call>,
+		trader: &mut Config::Trader,
+	) -> Result<Weight, XcmError> {
+		let mut total_surplus = 0;
 		match effect {
 			Order::DepositAsset { assets, dest } => {
 				let deposited = holding.saturating_take(assets);
 				for asset in deposited.into_assets_iter() {
 					Config::AssetTransactor::deposit_asset(&asset, &dest)?;
 				}
-				Ok(())
 			},
 			Order::DepositReserveAsset { assets, dest, effects } => {
 				let deposited = holding.saturating_take(assets);
 				for asset in deposited.assets_iter() {
 					Config::AssetTransactor::deposit_asset(&asset, &dest)?;
 				}
 				let assets = Self::reanchored(deposited, &dest);
-				Config::XcmSender::send_xcm(dest, Xcm::ReserveAssetDeposit { assets, effects })
+				Config::XcmSender::send_xcm(dest, Xcm::ReserveAssetDeposit { assets, effects })?;
 			},
 			Order::InitiateReserveWithdraw { assets, reserve, effects} => {
 				let assets = Self::reanchored(holding.saturating_take(assets), &reserve);
-				Config::XcmSender::send_xcm(reserve, Xcm::WithdrawAsset { assets, effects })
+				Config::XcmSender::send_xcm(reserve, Xcm::WithdrawAsset { assets, effects })?;
 			}
 			Order::InitiateTeleport { assets, dest, effects} => {
 				let assets = Self::reanchored(holding.saturating_take(assets), &dest);
-				Config::XcmSender::send_xcm(dest, Xcm::TeleportAsset { assets, effects })
+				Config::XcmSender::send_xcm(dest, Xcm::TeleportAsset { assets, effects })?;
 			}
 			Order::QueryHolding { query_id, dest, assets } => {
 				let assets = Self::reanchored(holding.min(assets.iter()), &dest);
-				Config::XcmSender::send_xcm(dest, Xcm::Balances { query_id, assets })
+				Config::XcmSender::send_xcm(dest, Xcm::QueryResponse { query_id, response: Response::Assets(assets) })?;
+			}
+			Order::BuyExecution { fees, weight, debt, halt_on_error, xcm } => {
+				// pay for `weight` using up to `fees` of the holding account.
+				let purchasing_weight = Weight::from(weight + debt);
```

Comment:
```suggestion
				let purchasing_weight = Weight::from(weight.saturating_add(debt));
```

Seems that `weight` and `debt` are controlled by the sender of this order.